fml/message_loop_task_queues.cc

```diff
@@ -91,196 +101,207 @@ void MessageLoopTaskQueues::GetTasksToRunNow(
-  task_observers_[queue_id].erase(key);
-}
+  std::scoped_lock queue_lock(GetMutex(queue_id));
 
-void MessageLoopTaskQueues::NotifyObservers(TaskQueueId queue_id) {
-  MergedQueuesRunner merged_observers =
-      MergedQueuesRunner(*this, queue_id, MutexType::kObservers);
-
-  merged_observers.InvokeMerged([&](TaskQueueId queue) {
-    for (const auto& observer : task_observers_[queue]) {
-      observer.second();
-    }
-  });
+  queue_entries_[queue_id].task_observers.erase(key);
 }
 
-// Thread safety analysis disabled as it does not account for defered locks.
-void MessageLoopTaskQueues::Swap(TaskQueueId primary, TaskQueueId secondary)
-    FML_NO_THREAD_SAFETY_ANALYSIS {
-  // task_observers locks
-  std::mutex& o1 = GetMutex(primary, MutexType::kObservers);
-  std::mutex& o2 = GetMutex(secondary, MutexType::kObservers);
+void MessageLoopTaskQueues::GetObserversToNotify(
+    TaskQueueId queue_id,
+    std::vector<fml::closure>& callables) const {
+  std::scoped_lock queue_lock(GetMutex(queue_id));
 
-  // delayed_tasks locks
-  std::mutex& t1 = GetMutex(primary, MutexType::kTasks);
-  std::mutex& t2 = GetMutex(secondary, MutexType::kTasks);
+  if (queue_entries_.at(queue_id).subsumed_by != _kUnmerged) {
+    return;
+  }
 
-  std::scoped_lock lock(o1, o2, t1, t2);
+  for (const auto& observer : queue_entries_.at(queue_id).task_observers) {
+    callables.push_back(observer.second);
+  }
 
-  std::swap(task_observers_[primary], task_observers_[secondary]);
-  std::swap(delayed_tasks_[primary], delayed_tasks_[secondary]);
+  TaskQueueId subsumed = queue_entries_.at(queue_id).owner_of;
+  if (subsumed != _kUnmerged) {
+    std::scoped_lock subsumed_lock(*queue_locks_.at(subsumed));
+    for (const auto& observer : queue_entries_.at(subsumed).task_observers) {
+      callables.push_back(observer.second);
+    }
+  }
 }
 
 void MessageLoopTaskQueues::SetWakeable(TaskQueueId queue_id,
                                         fml::Wakeable* wakeable) {
-  std::scoped_lock lock(GetMutex(queue_id, MutexType::kWakeables));
-  FML_CHECK(!wakeables_[queue_id]) << "Wakeable can only be set once.";
-  wakeables_[queue_id] = wakeable;
+  std::scoped_lock queue_lock(GetMutex(queue_id));
+
+  FML_CHECK(!queue_entries_[queue_id].wakeable)
+      << "Wakeable can only be set once.";
+  queue_entries_.at(queue_id).wakeable = wakeable;
 }
 
 bool MessageLoopTaskQueues::Merge(TaskQueueId owner, TaskQueueId subsumed) {
-  // task_observers locks
-  std::mutex& o1 = GetMutex(owner, MutexType::kObservers);
-  std::mutex& o2 = GetMutex(subsumed, MutexType::kObservers);
-
-  // delayed_tasks locks
-  std::mutex& t1 = GetMutex(owner, MutexType::kTasks);
-  std::mutex& t2 = GetMutex(subsumed, MutexType::kTasks);
-
-  std::scoped_lock lock(o1, o2, t1, t2);
-
   if (owner == subsumed) {
     return true;
   }
 
-  if (owner_to_subsumed_[owner] == subsumed) {
+  std::mutex& owner_lock = GetMutex(owner);
```

Comment:
ivar should probably be named mutex instead of lock. Also, `auto&` would be fine too.